doc/tutorials/ensemble.md

```diff
@@ -0,0 +1,465 @@
+of the module; it takes a port, starts an httpd server and waits for
+it to shutdown.  Notice that the server is initialized with the
+default mux, and it is initially empty; there are no handlers, so
+everything will 404.
+
+We also have a server wrapper for running the server as a service with `gxensemble run`.
+This is the code for the wrapper (see [src/tutorial/ensemble/httpd-svc.ss](https://github.com/vyzo/gerbil/tree/master/src/tutorial/ensemble/httpd-svc.ss)):
+```scheme
+(import ./server)
+(export main)
+
+(def (main (port "8080"))
+  (run! (string->number port)))
+```
+
+With all this, let's start an ensemble with two httpds, named `httpd1` and `httpd2`:
+```
+# generate a cookie for our ensemble, if on does not already exist
+$ gxensemble cookie
+
+# start the registry
+$ gxensemble registry
+...
+
+# start the servers
+$ gxensemble run --roles "(httpd)" httpd1 :tutorial/ensemble/httpd-svc 8080
+...
+
+$ gxensemble run --roles "(httpd)" httpd2 :tutorial/ensemble/httpd-svc 8081
+...
+```
+
+Now let's look at our servers:
+```
+$ gxensemble lookup --role httpd
+(httpd1 (unix dellicious /tmp/ensemble/httpd1))
+(httpd2 (unix dellicious /tmp/ensemble/httpd2))
+```
+
+We can also ping them for liveness:
+```
+$ gxensemble ping httpd1
+OK
+$ gxensemble ping httpd2
+OK
+```
+
+Now if we do a request at them we will see that there are no handlers:
+```
+$ curl -I http://localhost:8080/
+HTTP/1.1 404 Not Found
+Date: Tue Aug 15 06:34:30 2023
+Content-Length: 0
+
+$ curl -I http://localhost:8081/
+HTTP/1.1 404 Not Found
+Date: Tue Aug 15 06:34:43 2023
+Content-Length: 0
+```
+
+At the next step, we can load some code that implements handlers for our httpds;
+the code with the handlers lives in [src/tutorial/ensemble/handler.ss](https://github.com/vyzo/gerbil/tree/master/src/tutorial/ensemble/handler.ss):
+```scheme
+(import :std/net/httpd)
+(export #t)
+
+(def greeting
+  "hello there!\n")
+
+(def (set-greeting! what)
+  (set! greeting what))
+
+(def (write-simple-handler req res)
+  (http-response-write res 200 '(("Content-Type" . "text/plain"))
+                       greeting))
+
+(def (write-chunked-handler req res)
+  (http-response-begin res 200 '(("Content-Type" . "text/plain")))
+  (http-response-chunk res "hello ")
+  (http-response-chunk res "there!\n")
+  (http-response-end res))
+
+(def (root-handler req res)
+  (http-response-write res 200 [] "the world is not flat but round!\n"))
+
+(http-register-handler (current-http-server) "/" root-handler)
+```
+
+This code registers a root handler, and provides two more handlers
+that are not initially registerd anywhere.
+
+Here is how we can load the code:
+```
+# load with -f as there is no need to load any library dependencies
+$ gxensemble load -f httpd1 :tutorial/ensemble/handler
+... loading code object file /home/vyzo/.gerbil/lib/tutorial/ensemble/handler__0.o3
+ca3f193373a296d7bdb9101e7d4b9f1d450676aec6c49f05202a3dbcc5d766e2
```

Comment:
it will ship the object file, no need to build on the remote server.

Reply:
That's worthy of mention.

All of my questions are actually requests for clarification in the documentation rather than on github.

Reply:
Ok, I can add something in the section about loading.

> All of my questions are actually requests for clarification in the documentation rather than on github.

I understand that, and in generally I fix them in the documentation. But sometimes it needs to be discussed first, and some other times we should simply not say anything and avoid the "too much detail" problem that is pervasive in some of the docs you write :)
Don't get me wrong, I _like_ your exhaustive documentation style but sometimes it is just too much and I just TL;DR.

Also you shouldn't expect me to write exhaustive documentation; that takes time and there is code to write.
You can help out by reading the code and improving the documentation!